q2_mlab/plotting/app.py

```diff
@@ -0,0 +1,613 @@
+        if (event_name == 'selected-indices') and \
+                (component is self.scatter_source):
+            selected_indices = self.scatter_source.data_source.selected.indices
+            self.dataset_bars_source.data = self.get_dataset_counts(
+                indices=selected_indices,
+            )
+            self.level_bars_source.data = self.get_level_counts(
+                indices=selected_indices,
+            )
+            self.target_bars_source.data = self.get_target_counts(
+                indices=selected_indices,
+            )
+        if (event_name == 'button-click') and \
+                (component is self.query_button):
+            df = self.handle_query(self.query_input.text_input.value)
+            # need to update self.data due to how the hbars are currently
+            # written
+            self.data = df
+            self.scatter_source.data_source.data = df.to_dict(orient='list')
+            segment_source = find_segments(
+                df,
+                across=self.line_segment_variable,
+                groupby=['parameters_id', 'algorithm', 'level', 'dataset',
+                         'target'],
+            )
+            self.segment.segment.data_source.data = segment_source.to_dict(
+                orient='list',
+            )
+
+            selected_indices = self.scatter_source.data_source.selected.indices
+            self.dataset_bars_source.data = self.get_dataset_counts(
+                indices=selected_indices,
+            )
+            self.level_bars_source.data = self.get_level_counts(
+                indices=selected_indices,
+            )
+            self.target_bars_source.data = self.get_target_counts(
+                indices=selected_indices,
+            )
+        if (event_name == 'checkbox-change') and \
+                (component is self.segment_button):
+            active = self.segment_button.checkbox.active
+            if 0 in active:
+                self.segment.segment.visible = True
+            else:
+                self.segment.segment.visible = False
+        if (event_name == 'dropdown-select') and \
+                (component is self.segment_variable_select):
+            new_segment_variable = self.segment_variable_select.select.value
+            self.line_segment_variable = new_segment_variable
+            new_segment_data = find_segments(
+                self.data,
+                across=self.line_segment_variable,
+                groupby=['parameters_id', 'algorithm', 'level', 'dataset',
+                         'target']
+            )
+            line_segment_ends = self.line_segment_pairs[new_segment_variable]
+            self.segment.redraw(
+                self.x,
+                self.y,
+                *line_segment_ends,
+                new_segment_data
+            )
+
+    def plot(self):
+        self.data_raw = pd.read_sql_table(RegressionScore.__tablename__,
```

Comment:
Similarly for classification - the table we read should probably be given as a cli argument (+ some other changes to accommodate classification results). In scope, or should we save it for the next PR?

Reply:
Logic here might get complex. Could warrant its own PR as this one is already pretty huge.